
# **Learning Pandas**


---



### Introduction
Just like we imported numpy, we have to import pandas aswell.

In [1]:
import numpy as np
import pandas as pd

In similar fashion we can check the version as well

In [2]:
pd.__version__

'1.5.3'

We'll use the following data set to perform further operations

In [3]:
# List of cloud technologies in 2023
cloud_2023 = ['Amazon Web Services (AWS)', 'Firebase', 'Cloudflare', 'Google Cloud',
              'Digital Ocean','Heroku', 'Vercel', 'Netlify', 'VMware', 'Hetzner',
              'Linode, now Akamai', 'Managed Hosting', 'OVH','Oracle Cloud Infrastructure (OCI)',
              'OpenShift', 'Fly.io', 'Vultr', 'Render', 'OpenStack','IBM Cloud Or Watson', 'Scaleway', 'Colocation', 'Microsoft Azure']



*   The Series property is used to tabularize one dimensional data in pandas
*   Let's convert this list into pandas object using series:-





In [4]:
series_obj = pd.Series(cloud_2023) # It does it according to the indexing
print("This is the pandas object:\n")
print(series_obj,"\n\nThe data type is",type(series_obj))

This is the pandas object:

0             Amazon Web Services (AWS)
1                              Firebase
2                            Cloudflare
3                          Google Cloud
4                         Digital Ocean
5                                Heroku
6                                Vercel
7                               Netlify
8                                VMware
9                               Hetzner
10                   Linode, now Akamai
11                      Managed Hosting
12                                  OVH
13    Oracle Cloud Infrastructure (OCI)
14                            OpenShift
15                               Fly.io
16                                Vultr
17                               Render
18                            OpenStack
19                  IBM Cloud Or Watson
20                             Scaleway
21                           Colocation
22                      Microsoft Azure
dtype: object 

The data type is <class 'pandas.core

The property **index** will give idea about the pandas object's index

In [5]:
series_obj.index

RangeIndex(start=0, stop=23, step=1)

We can also convert index details into numpy array

In [6]:
indexes_array = np.array(series_obj.index)
indexes_array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22])

Similarly we can have the array of the values using the **values** property

In [7]:
values_array = series_obj.values
print(f"The type of values_array is {type(values_array)} and the values are:\n\n",values_array)

The type of values_array is <class 'numpy.ndarray'> and the values are:

 ['Amazon Web Services (AWS)' 'Firebase' 'Cloudflare' 'Google Cloud'
 'Digital Ocean' 'Heroku' 'Vercel' 'Netlify' 'VMware' 'Hetzner'
 'Linode, now Akamai' 'Managed Hosting' 'OVH'
 'Oracle Cloud Infrastructure (OCI)' 'OpenShift' 'Fly.io' 'Vultr' 'Render'
 'OpenStack' 'IBM Cloud Or Watson' 'Scaleway' 'Colocation'
 'Microsoft Azure']


The reason we got the values as an array because the pandas framework in build on numpy.



### Slicing
As pandas is built on numpy, same concepts are applicable for slicing

In [8]:
print("The first element of the array is",series_obj[0])
print("The next four elements of the array are")
print(series_obj[1:5])

The first element of the array is Amazon Web Services (AWS)
The next four elements of the array are
1         Firebase
2       Cloudflare
3     Google Cloud
4    Digital Ocean
dtype: object


We'll use the same method to reverse the object

In [9]:
print("The reversed series_obj is as follows\n")
print(series_obj[::-1])

The reversed series_obj is as follows

22                      Microsoft Azure
21                           Colocation
20                             Scaleway
19                  IBM Cloud Or Watson
18                            OpenStack
17                               Render
16                                Vultr
15                               Fly.io
14                            OpenShift
13    Oracle Cloud Infrastructure (OCI)
12                                  OVH
11                      Managed Hosting
10                   Linode, now Akamai
9                               Hetzner
8                                VMware
7                               Netlify
6                                Vercel
5                                Heroku
4                         Digital Ocean
3                          Google Cloud
2                            Cloudflare
1                              Firebase
0             Amazon Web Services (AWS)
dtype: object


### Custom indexing

Unlike numpy where we can have only numeric indexes, in pandas we can have **anything** as index

In [10]:
series_obj_custom = pd.Series(['Amazon Web Services (AWS)', 'Firebase', 'Cloudflare', 'Google Cloud',
              'Digital Ocean'],index=['a','b','c','d','e'])
print("The series with custom indexing:-")
print(series_obj_custom)

The series with custom indexing:-
a    Amazon Web Services (AWS)
b                     Firebase
c                   Cloudflare
d                 Google Cloud
e                Digital Ocean
dtype: object


### Slicing using Custom indexes
We can also perform slicing and stuff using our custom indexes

In [11]:
print("The series from 'a' to 'd' is as follows")
print(series_obj_custom['a':'e'])

The series from 'a' to 'd' is as follows
a    Amazon Web Services (AWS)
b                     Firebase
c                   Cloudflare
d                 Google Cloud
e                Digital Ocean
dtype: object


**Note**: During slicing with custom indexes we'll have both the limits included

### Data Set

We'll use the following data set to perform further operations

In [12]:
# List of cloud technologies in 2023
cloud_2023 = ['Amazon Web Services (AWS)', 'Microsoft Azure', 'Google Cloud', 'Firebase', 'Cloudflare', 'Digital Ocean',
               'Heroku', 'Vercel', 'Netlify', 'VMware', 'Hetzner', 'Linode, now Akamai', 'Managed Hosting', 'OVH',
               'Oracle Cloud Infrastructure (OCI)', 'OpenShift', 'Fly.io', 'Vultr', 'Render', 'OpenStack',
               'IBM Cloud Or Watson', 'Scaleway', 'Colocation']

# List of total votes for each cloud technology
total_votes = [33818, 18105, 16592, 10761, 10599, 9299, 8358, 7431, 6228, 4964, 3069, 2755, 2379, 2378, 1810,
               1671, 1649, 1357, 1287, 1076, 801, 628, 497]

# List of professional votes for each cloud technology
professional_votes = [29479, 15441, 13304, 8550, 8326, 7838, 6535, 5727, 4708, 3584, 2465, 2104, 1889, 1912,
                      1146, 1432, 1342, 1021, 920, 802, 571, 520, 385]

# List of beginner votes for each cloud technology
beginner_votes = [654, 370, 652, 628, 420, 247, 492, 487, 498, 262, 67, 125, 63, 76, 123, 22, 72, 55,
                  132, 53, 53, 0, 20]

# List of other votes for each cloud technology
other_votes = [1185, 714, 866, 486, 621, 393, 418, 392, 330, 396, 184, 167, 137, 126, 190, 66, 79,
                94, 69, 78, 70, 31, 30]

### Using DataFrame

First we'll have a serial number array using numpy's built in function called **np.arrange(start, end, step)**

Note: The end value is exclusive

In [13]:
sno = np.arange(1,len(cloud_2023)+1) # To make a numpy array of the given range

Now we'll create a data frame to tabularize our data.

We can have two methods to do this:

***Method-1 (Square Bracket Notation)***

-> We'll add colums like array notaion

In [14]:
cloud_survey_table = pd.DataFrame()
cloud_survey_table["SNO"] = sno
cloud_survey_table["CLOUD_PLATFORM"] = cloud_2023
cloud_survey_table["TOTAL_VOTES"] = total_votes
cloud_survey_table["PROFESSIONAL_VOTES"] = professional_votes
cloud_survey_table["BEGINNER_VOTES"] = beginner_votes
cloud_survey_table["OTHER_VOTES"] = other_votes

print("The Cloud Survey table is as follows\n")
cloud_survey_table

The Cloud Survey table is as follows



SNO                     CLOUD_PLATFORM  TOTAL_VOTES  PROFESSIONAL_VOTES  \
0     1          Amazon Web Services (AWS)        33818               29479   
1     2                    Microsoft Azure        18105               15441   
2     3                       Google Cloud        16592               13304   
3     4                           Firebase        10761                8550   
4     5                         Cloudflare        10599                8326   
5     6                      Digital Ocean         9299                7838   
6     7                             Heroku         8358                6535   
7     8                             Vercel         7431                5727   
8     9                            Netlify         6228                4708   
9    10                             VMware         4964                3584   
10   11                            Hetzner         3069                2465   
11   12                 Linode, now Akamai         2755                2104   
12   13                    Managed Hosting         2379                1889   
13   14                                OVH         2378                1912   
14   15  Oracle Cloud Infrastructure (OCI)         1810                1146   
15   16                          OpenShift         1671                1432   
16   17                             Fly.io         1649                1342   
17   18                              Vultr         1357                1021   
18   19                             Render         1287                 920   
19   20                          OpenStack         1076                 802   
20   21                IBM Cloud Or Watson          801                 571   
21   22                           Scaleway          628                 520   
22   23                         Colocation          497                 385   

    BEGINNER_VOTES  OTHER_VOTES  
0              654         1185  
1              370          714  
2              652          866  
3              628          486  
4              420          621  
5              247          393  
6              492          418  
7              487          392  
8              498          330  
9              262          396  
10              67          184  
11             125          167  
12              63          137  
13              76          126  
14             123          190  
15              22           66  
16              72           79  
17              55           94  
18             132           69  
19              53           78  
20              53           70  
21               0           31  
22              20           30

We can access the columns using columns property

In [15]:
cloud_survey_table.index

RangeIndex(start=0, stop=23, step=1)

In [16]:
cloud_survey_table.columns

Index(['SNO', 'CLOUD_PLATFORM', 'TOTAL_VOTES', 'PROFESSIONAL_VOTES',
       'BEGINNER_VOTES', 'OTHER_VOTES'],
      dtype='object')

***Method-2 (Using dictionary)***

In [17]:
cloud_survey_table = pd.DataFrame({"CLOUD_PLATFORMS":cloud_2023,
                                   "TOTAL_VOTES":total_votes,
                                   "PROFESSIONAL_VOTES":professional_votes,
                                   "BEGINNER_VOTES":beginner_votes,
                                   "OTHER_VOTES":other_votes,
                                   },index=sno)
# we have to use dictionary only for the column names and column data
cloud_survey_table

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
1           Amazon Web Services (AWS)        33818               29479   
2                     Microsoft Azure        18105               15441   
3                        Google Cloud        16592               13304   
4                            Firebase        10761                8550   
5                          Cloudflare        10599                8326   
6                       Digital Ocean         9299                7838   
7                              Heroku         8358                6535   
8                              Vercel         7431                5727   
9                             Netlify         6228                4708   
10                             VMware         4964                3584   
11                            Hetzner         3069                2465   
12                 Linode, now Akamai         2755                2104   
13                    Managed Hosting         2379                1889   
14                                OVH         2378                1912   
15  Oracle Cloud Infrastructure (OCI)         1810                1146   
16                          OpenShift         1671                1432   
17                             Fly.io         1649                1342   
18                              Vultr         1357                1021   
19                             Render         1287                 920   
20                          OpenStack         1076                 802   
21                IBM Cloud Or Watson          801                 571   
22                           Scaleway          628                 520   
23                         Colocation          497                 385   

    BEGINNER_VOTES  OTHER_VOTES  
1              654         1185  
2              370          714  
3              652          866  
4              628          486  
5              420          621  
6              247          393  
7              492          418  
8              487          392  
9              498          330  
10             262          396  
11              67          184  
12             125          167  
13              63          137  
14              76          126  
15             123          190  
16              22           66  
17              72           79  
18              55           94  
19             132           69  
20              53           78  
21              53           70  
22               0           31  
23              20           30

This type of syntax is usually preferred as it's a bit easier and cleaner

### Implicit and Explicit indexing

In [18]:
series_obj = pd.Series(['a','b','c','d'],index=[1,2,3,4])
series_obj

1    a
2    b
3    c
4    d
dtype: object

There is a confusion regarding indexes when we give a list of numbers as a custom index. To avoid this we'll use loc (explicit) i.e. to use default pandas indexing and iloc (implicit) i.e. to use our own custom indexing

In [19]:
print("This is explicit result =",series_obj.loc[1])
print("This is implicit result =",series_obj.iloc[1])

This is explicit result = a
This is implicit result = b


### Accessing particular columns

We have two methods for this

**Method-1**

Using Brackets notation

In [20]:
cloud_survey_table["CLOUD_PLATFORMS"]

1             Amazon Web Services (AWS)
2                       Microsoft Azure
3                          Google Cloud
4                              Firebase
5                            Cloudflare
6                         Digital Ocean
7                                Heroku
8                                Vercel
9                               Netlify
10                               VMware
11                              Hetzner
12                   Linode, now Akamai
13                      Managed Hosting
14                                  OVH
15    Oracle Cloud Infrastructure (OCI)
16                            OpenShift
17                               Fly.io
18                                Vultr
19                               Render
20                            OpenStack
21                  IBM Cloud Or Watson
22                             Scaleway
23                           Colocation
Name: CLOUD_PLATFORMS, dtype: object

**Method-2**

Using Dot notation

*This method is preferred as we can club other numpy properties with it*

In [21]:
cloud_survey_table.CLOUD_PLATFORMS

1             Amazon Web Services (AWS)
2                       Microsoft Azure
3                          Google Cloud
4                              Firebase
5                            Cloudflare
6                         Digital Ocean
7                                Heroku
8                                Vercel
9                               Netlify
10                               VMware
11                              Hetzner
12                   Linode, now Akamai
13                      Managed Hosting
14                                  OVH
15    Oracle Cloud Infrastructure (OCI)
16                            OpenShift
17                               Fly.io
18                                Vultr
19                               Render
20                            OpenStack
21                  IBM Cloud Or Watson
22                             Scaleway
23                           Colocation
Name: CLOUD_PLATFORMS, dtype: object

We can also club things using method-1 but not all the properties work with it!

But method-1 has it's own advantage too i.e. it can be used to print multiple colummns

In [22]:
cloud_survey_table[["CLOUD_PLATFORMS","TOTAL_VOTES"]] # We have to use list

CLOUD_PLATFORMS  TOTAL_VOTES
1           Amazon Web Services (AWS)        33818
2                     Microsoft Azure        18105
3                        Google Cloud        16592
4                            Firebase        10761
5                          Cloudflare        10599
6                       Digital Ocean         9299
7                              Heroku         8358
8                              Vercel         7431
9                             Netlify         6228
10                             VMware         4964
11                            Hetzner         3069
12                 Linode, now Akamai         2755
13                    Managed Hosting         2379
14                                OVH         2378
15  Oracle Cloud Infrastructure (OCI)         1810
16                          OpenShift         1671
17                             Fly.io         1649
18                              Vultr         1357
19                             Render         1287
20                          OpenStack         1076
21                IBM Cloud Or Watson          801
22                           Scaleway          628
23                         Colocation          497

We can use the .T property to transpose the table

### Microactions

In [23]:
cloud_survey_table.T

1                2             3   \
CLOUD_PLATFORMS     Amazon Web Services (AWS)  Microsoft Azure  Google Cloud   
TOTAL_VOTES                             33818            18105         16592   
PROFESSIONAL_VOTES                      29479            15441         13304   
BEGINNER_VOTES                            654              370           652   
OTHER_VOTES                              1185              714           866   

                          4           5              6       7       8   \
CLOUD_PLATFORMS     Firebase  Cloudflare  Digital Ocean  Heroku  Vercel   
TOTAL_VOTES            10761       10599           9299    8358    7431   
PROFESSIONAL_VOTES      8550        8326           7838    6535    5727   
BEGINNER_VOTES           628         420            247     492     487   
OTHER_VOTES              486         621            393     418     392   

                         9       10  ...    14  \
CLOUD_PLATFORMS     Netlify  VMware  ...   OVH   
TOTAL_VOTES            6228    4964  ...  2378   
PROFESSIONAL_VOTES     4708    3584  ...  1912   
BEGINNER_VOTES          498     262  ...    76   
OTHER_VOTES             330     396  ...   126   

                                                   15         16      17  \
CLOUD_PLATFORMS     Oracle Cloud Infrastructure (OCI)  OpenShift  Fly.io   
TOTAL_VOTES                                      1810       1671    1649   
PROFESSIONAL_VOTES                               1146       1432    1342   
BEGINNER_VOTES                                    123         22      72   
OTHER_VOTES                                       190         66      79   

                       18      19         20                   21        22  \
CLOUD_PLATFORMS     Vultr  Render  OpenStack  IBM Cloud Or Watson  Scaleway   
TOTAL_VOTES          1357    1287       1076                  801       628   
PROFESSIONAL_VOTES   1021     920        802                  571       520   
BEGINNER_VOTES         55     132         53                   53         0   
OTHER_VOTES            94      69         78                   70        31   

                            23  
CLOUD_PLATFORMS     Colocation  
TOTAL_VOTES                497  
PROFESSIONAL_VOTES         385  
BEGINNER_VOTES              20  
OTHER_VOTES                 30  

[5 rows x 23 columns]

We can also perform certain actions like average, percentage etc and add it to the dataframe

In [24]:
cloud_survey_table["PERCENTAGE"] = round((cloud_survey_table["TOTAL_VOTES"] / np.add.reduce(cloud_survey_table["TOTAL_VOTES"]))*100,2)
print("Percentage votes of each cloud platform")
cloud_survey_table.PERCENTAGE

Percentage votes of each cloud platform


1     22.93
2     12.27
3     11.25
4      7.29
5      7.19
6      6.30
7      5.67
8      5.04
9      4.22
10     3.37
11     2.08
12     1.87
13     1.61
14     1.61
15     1.23
16     1.13
17     1.12
18     0.92
19     0.87
20     0.73
21     0.54
22     0.43
23     0.34
Name: PERCENTAGE, dtype: float64

In [25]:
cloud_survey_table

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
1           Amazon Web Services (AWS)        33818               29479   
2                     Microsoft Azure        18105               15441   
3                        Google Cloud        16592               13304   
4                            Firebase        10761                8550   
5                          Cloudflare        10599                8326   
6                       Digital Ocean         9299                7838   
7                              Heroku         8358                6535   
8                              Vercel         7431                5727   
9                             Netlify         6228                4708   
10                             VMware         4964                3584   
11                            Hetzner         3069                2465   
12                 Linode, now Akamai         2755                2104   
13                    Managed Hosting         2379                1889   
14                                OVH         2378                1912   
15  Oracle Cloud Infrastructure (OCI)         1810                1146   
16                          OpenShift         1671                1432   
17                             Fly.io         1649                1342   
18                              Vultr         1357                1021   
19                             Render         1287                 920   
20                          OpenStack         1076                 802   
21                IBM Cloud Or Watson          801                 571   
22                           Scaleway          628                 520   
23                         Colocation          497                 385   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  
1              654         1185       22.93  
2              370          714       12.27  
3              652          866       11.25  
4              628          486        7.29  
5              420          621        7.19  
6              247          393        6.30  
7              492          418        5.67  
8              487          392        5.04  
9              498          330        4.22  
10             262          396        3.37  
11              67          184        2.08  
12             125          167        1.87  
13              63          137        1.61  
14              76          126        1.61  
15             123          190        1.23  
16              22           66        1.13  
17              72           79        1.12  
18              55           94        0.92  
19             132           69        0.87  
20              53           78        0.73  
21              53           70        0.54  
22               0           31        0.43  
23              20           30        0.34

We can use the copy method to copy one pandas object into another

In [26]:
cloud_survey_table_missing = cloud_survey_table.copy()

### Filling, altering and dropping values

*   We can also edit the values of a particular place using the following syntax
*   To represent Not a number we'll use np.nan



In [27]:
cloud_survey_table_missing.loc[1,"PROFESSIONAL_VOTES"] = np.nan
cloud_survey_table_missing

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
1           Amazon Web Services (AWS)        33818                 NaN   
2                     Microsoft Azure        18105             15441.0   
3                        Google Cloud        16592             13304.0   
4                            Firebase        10761              8550.0   
5                          Cloudflare        10599              8326.0   
6                       Digital Ocean         9299              7838.0   
7                              Heroku         8358              6535.0   
8                              Vercel         7431              5727.0   
9                             Netlify         6228              4708.0   
10                             VMware         4964              3584.0   
11                            Hetzner         3069              2465.0   
12                 Linode, now Akamai         2755              2104.0   
13                    Managed Hosting         2379              1889.0   
14                                OVH         2378              1912.0   
15  Oracle Cloud Infrastructure (OCI)         1810              1146.0   
16                          OpenShift         1671              1432.0   
17                             Fly.io         1649              1342.0   
18                              Vultr         1357              1021.0   
19                             Render         1287               920.0   
20                          OpenStack         1076               802.0   
21                IBM Cloud Or Watson          801               571.0   
22                           Scaleway          628               520.0   
23                         Colocation          497               385.0   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  
1              654         1185       22.93  
2              370          714       12.27  
3              652          866       11.25  
4              628          486        7.29  
5              420          621        7.19  
6              247          393        6.30  
7              492          418        5.67  
8              487          392        5.04  
9              498          330        4.22  
10             262          396        3.37  
11              67          184        2.08  
12             125          167        1.87  
13              63          137        1.61  
14              76          126        1.61  
15             123          190        1.23  
16              22           66        1.13  
17              72           79        1.12  
18              55           94        0.92  
19             132           69        0.87  
20              53           78        0.73  
21              53           70        0.54  
22               0           31        0.43  
23              20           30        0.34

When we use NaN value at any one place it changes the datatype of the element to float

In [28]:
cloud_survey_table_missing.loc[3,"BEGINNER_VOTES"] = np.nan
cloud_survey_table_missing.loc[4,"OTHER_VOTES"] = np.nan
cloud_survey_table_missing

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
1           Amazon Web Services (AWS)        33818                 NaN   
2                     Microsoft Azure        18105             15441.0   
3                        Google Cloud        16592             13304.0   
4                            Firebase        10761              8550.0   
5                          Cloudflare        10599              8326.0   
6                       Digital Ocean         9299              7838.0   
7                              Heroku         8358              6535.0   
8                              Vercel         7431              5727.0   
9                             Netlify         6228              4708.0   
10                             VMware         4964              3584.0   
11                            Hetzner         3069              2465.0   
12                 Linode, now Akamai         2755              2104.0   
13                    Managed Hosting         2379              1889.0   
14                                OVH         2378              1912.0   
15  Oracle Cloud Infrastructure (OCI)         1810              1146.0   
16                          OpenShift         1671              1432.0   
17                             Fly.io         1649              1342.0   
18                              Vultr         1357              1021.0   
19                             Render         1287               920.0   
20                          OpenStack         1076               802.0   
21                IBM Cloud Or Watson          801               571.0   
22                           Scaleway          628               520.0   
23                         Colocation          497               385.0   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  
1            654.0       1185.0       22.93  
2            370.0        714.0       12.27  
3              NaN        866.0       11.25  
4            628.0          NaN        7.29  
5            420.0        621.0        7.19  
6            247.0        393.0        6.30  
7            492.0        418.0        5.67  
8            487.0        392.0        5.04  
9            498.0        330.0        4.22  
10           262.0        396.0        3.37  
11            67.0        184.0        2.08  
12           125.0        167.0        1.87  
13            63.0        137.0        1.61  
14            76.0        126.0        1.61  
15           123.0        190.0        1.23  
16            22.0         66.0        1.13  
17            72.0         79.0        1.12  
18            55.0         94.0        0.92  
19           132.0         69.0        0.87  
20            53.0         78.0        0.73  
21            53.0         70.0        0.54  
22             0.0         31.0        0.43  
23            20.0         30.0        0.34

We can also detect null values by using isnull function...It returns true at the places where the value is null and false everywhere else

In [29]:
cloud_survey_table_missing.isnull()

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  BEGINNER_VOTES  \
1             False        False                True           False   
2             False        False               False           False   
3             False        False               False            True   
4             False        False               False           False   
5             False        False               False           False   
6             False        False               False           False   
7             False        False               False           False   
8             False        False               False           False   
9             False        False               False           False   
10            False        False               False           False   
11            False        False               False           False   
12            False        False               False           False   
13            False        False               False           False   
14            False        False               False           False   
15            False        False               False           False   
16            False        False               False           False   
17            False        False               False           False   
18            False        False               False           False   
19            False        False               False           False   
20            False        False               False           False   
21            False        False               False           False   
22            False        False               False           False   
23            False        False               False           False   

    OTHER_VOTES  PERCENTAGE  
1         False       False  
2         False       False  
3         False       False  
4          True       False  
5         False       False  
6         False       False  
7         False       False  
8         False       False  
9         False       False  
10        False       False  
11        False       False  
12        False       False  
13        False       False  
14        False       False  
15        False       False  
16        False       False  
17        False       False  
18        False       False  
19        False       False  
20        False       False  
21        False       False  
22        False       False  
23        False       False

We can either fill or drop (removing the whole row) NaN values





Let's see how to drop NaN values :-

In [30]:
cloud_survey_table_missing.dropna()

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
2                     Microsoft Azure        18105             15441.0   
5                          Cloudflare        10599              8326.0   
6                       Digital Ocean         9299              7838.0   
7                              Heroku         8358              6535.0   
8                              Vercel         7431              5727.0   
9                             Netlify         6228              4708.0   
10                             VMware         4964              3584.0   
11                            Hetzner         3069              2465.0   
12                 Linode, now Akamai         2755              2104.0   
13                    Managed Hosting         2379              1889.0   
14                                OVH         2378              1912.0   
15  Oracle Cloud Infrastructure (OCI)         1810              1146.0   
16                          OpenShift         1671              1432.0   
17                             Fly.io         1649              1342.0   
18                              Vultr         1357              1021.0   
19                             Render         1287               920.0   
20                          OpenStack         1076               802.0   
21                IBM Cloud Or Watson          801               571.0   
22                           Scaleway          628               520.0   
23                         Colocation          497               385.0   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  
2            370.0        714.0       12.27  
5            420.0        621.0        7.19  
6            247.0        393.0        6.30  
7            492.0        418.0        5.67  
8            487.0        392.0        5.04  
9            498.0        330.0        4.22  
10           262.0        396.0        3.37  
11            67.0        184.0        2.08  
12           125.0        167.0        1.87  
13            63.0        137.0        1.61  
14            76.0        126.0        1.61  
15           123.0        190.0        1.23  
16            22.0         66.0        1.13  
17            72.0         79.0        1.12  
18            55.0         94.0        0.92  
19           132.0         69.0        0.87  
20            53.0         78.0        0.73  
21            53.0         70.0        0.54  
22             0.0         31.0        0.43  
23            20.0         30.0        0.34

If we try to perform actions on array which have NaN values then the result will be NaN

In [31]:
print("Mean value without any change :-",np.add.reduce(cloud_survey_table_missing.PROFESSIONAL_VOTES)/len(cloud_survey_table_missing.PROFESSIONAL_VOTES))

Mean value without any change :- nan


*   We can fill the NaN values with other values
*   For that we'll use the fillna() and we'll pass a dictionary with column name as the key and the replacement value as the value

Let's start by calculating the mean after replacing NaN values with zero

In [32]:
cloud_survey_table_filled = cloud_survey_table_missing.fillna({"PROFESSIONAL_VOTES":0,"BEGINNER_VOTES":0,"OTHER_VOTES":0})
print("Mean value after filling NaN values with zero :-",round(np.add.reduce(cloud_survey_table_filled.PROFESSIONAL_VOTES)/len(cloud_survey_table_filled.PROFESSIONAL_VOTES),2))


Mean value after filling NaN values with zero :- 3935.74


Now, Let's try to calculate the mean by dropping the NaN values

Note: Both the value and number will decrease

In [33]:
print("Mean value after dropping NaN values :-",round(np.add.reduce(cloud_survey_table_missing.PROFESSIONAL_VOTES.dropna())/len(cloud_survey_table_missing.PROFESSIONAL_VOTES.dropna()),2))


Mean value after dropping NaN values :- 4114.64


Instead of calculating mean manually we can also use this numpy function named **mean()** or **nanmean()**



*   mean(): Used when the array doesn't have any NaN value
*   namean(): Used when the array does have a NaN value



In [34]:
print("Mean value after dropping NaN values using dropna() and mean():-",round(np.mean(cloud_survey_table_missing.PROFESSIONAL_VOTES.dropna()),2))
print("Mean value after dropping NaN values using array with NaN values and nanmean():-",round(np.nanmean(cloud_survey_table_missing['PROFESSIONAL_VOTES']),2))

Mean value after dropping NaN values using dropna() and mean():- 4114.64
Mean value after dropping NaN values using array with NaN values and nanmean():- 4114.64


Another method to fill the missing values is by using the value which is just next to it. We'll do so using the same method **fillna()** with the argument **method='ffill'**

In [35]:
cloud_survey_table_missing.fillna(method="ffill") # Forward Fill
# We usually use this approach in place like stock price

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
1           Amazon Web Services (AWS)        33818                 NaN   
2                     Microsoft Azure        18105             15441.0   
3                        Google Cloud        16592             13304.0   
4                            Firebase        10761              8550.0   
5                          Cloudflare        10599              8326.0   
6                       Digital Ocean         9299              7838.0   
7                              Heroku         8358              6535.0   
8                              Vercel         7431              5727.0   
9                             Netlify         6228              4708.0   
10                             VMware         4964              3584.0   
11                            Hetzner         3069              2465.0   
12                 Linode, now Akamai         2755              2104.0   
13                    Managed Hosting         2379              1889.0   
14                                OVH         2378              1912.0   
15  Oracle Cloud Infrastructure (OCI)         1810              1146.0   
16                          OpenShift         1671              1432.0   
17                             Fly.io         1649              1342.0   
18                              Vultr         1357              1021.0   
19                             Render         1287               920.0   
20                          OpenStack         1076               802.0   
21                IBM Cloud Or Watson          801               571.0   
22                           Scaleway          628               520.0   
23                         Colocation          497               385.0   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  
1            654.0       1185.0       22.93  
2            370.0        714.0       12.27  
3            370.0        866.0       11.25  
4            628.0        866.0        7.29  
5            420.0        621.0        7.19  
6            247.0        393.0        6.30  
7            492.0        418.0        5.67  
8            487.0        392.0        5.04  
9            498.0        330.0        4.22  
10           262.0        396.0        3.37  
11            67.0        184.0        2.08  
12           125.0        167.0        1.87  
13            63.0        137.0        1.61  
14            76.0        126.0        1.61  
15           123.0        190.0        1.23  
16            22.0         66.0        1.13  
17            72.0         79.0        1.12  
18            55.0         94.0        0.92  
19           132.0         69.0        0.87  
20            53.0         78.0        0.73  
21            53.0         70.0        0.54  
22             0.0         31.0        0.43  
23            20.0         30.0        0.34

We can also do this using the backward propagation

In [36]:
cloud_survey_table_missing.fillna(method="bfill") # Backward Fill

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
1           Amazon Web Services (AWS)        33818             15441.0   
2                     Microsoft Azure        18105             15441.0   
3                        Google Cloud        16592             13304.0   
4                            Firebase        10761              8550.0   
5                          Cloudflare        10599              8326.0   
6                       Digital Ocean         9299              7838.0   
7                              Heroku         8358              6535.0   
8                              Vercel         7431              5727.0   
9                             Netlify         6228              4708.0   
10                             VMware         4964              3584.0   
11                            Hetzner         3069              2465.0   
12                 Linode, now Akamai         2755              2104.0   
13                    Managed Hosting         2379              1889.0   
14                                OVH         2378              1912.0   
15  Oracle Cloud Infrastructure (OCI)         1810              1146.0   
16                          OpenShift         1671              1432.0   
17                             Fly.io         1649              1342.0   
18                              Vultr         1357              1021.0   
19                             Render         1287               920.0   
20                          OpenStack         1076               802.0   
21                IBM Cloud Or Watson          801               571.0   
22                           Scaleway          628               520.0   
23                         Colocation          497               385.0   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  
1            654.0       1185.0       22.93  
2            370.0        714.0       12.27  
3            628.0        866.0       11.25  
4            628.0        621.0        7.29  
5            420.0        621.0        7.19  
6            247.0        393.0        6.30  
7            492.0        418.0        5.67  
8            487.0        392.0        5.04  
9            498.0        330.0        4.22  
10           262.0        396.0        3.37  
11            67.0        184.0        2.08  
12           125.0        167.0        1.87  
13            63.0        137.0        1.61  
14            76.0        126.0        1.61  
15           123.0        190.0        1.23  
16            22.0         66.0        1.13  
17            72.0         79.0        1.12  
18            55.0         94.0        0.92  
19           132.0         69.0        0.87  
20            53.0         78.0        0.73  
21            53.0         70.0        0.54  
22             0.0         31.0        0.43  
23            20.0         30.0        0.34

### Concatenation
Now let's try to understand how we can add a row in a pandas table:

For this we'll first declare the row and then we'll add it to the table.

NOTE: Even if it's a single value we'll still use list and it should be present in a dictionary

In [37]:
new_cloud_table = pd.DataFrame({
    "CLOUD_PLATFORMS":["My Cloud Platform"],
    "TOTAL_VOTES":[500],
    "PROFESSIONAL_VOTES":[323],
    "BEGINNER_VOTES":[123],
    "OTHER_VOTES":[67]
    }
)
new_cloud_table

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  BEGINNER_VOTES  \
0  My Cloud Platform          500                 323             123   

   OTHER_VOTES  
0           67

Now let's add it to the table

For this we'll use `.concat([],ignore_index="true")`

In [38]:
updated_table = pd.concat([cloud_survey_table,new_cloud_table],ignore_index="True")
updated_table

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
0           Amazon Web Services (AWS)        33818               29479   
1                     Microsoft Azure        18105               15441   
2                        Google Cloud        16592               13304   
3                            Firebase        10761                8550   
4                          Cloudflare        10599                8326   
5                       Digital Ocean         9299                7838   
6                              Heroku         8358                6535   
7                              Vercel         7431                5727   
8                             Netlify         6228                4708   
9                              VMware         4964                3584   
10                            Hetzner         3069                2465   
11                 Linode, now Akamai         2755                2104   
12                    Managed Hosting         2379                1889   
13                                OVH         2378                1912   
14  Oracle Cloud Infrastructure (OCI)         1810                1146   
15                          OpenShift         1671                1432   
16                             Fly.io         1649                1342   
17                              Vultr         1357                1021   
18                             Render         1287                 920   
19                          OpenStack         1076                 802   
20                IBM Cloud Or Watson          801                 571   
21                           Scaleway          628                 520   
22                         Colocation          497                 385   
23                  My Cloud Platform          500                 323   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  
0              654         1185       22.93  
1              370          714       12.27  
2              652          866       11.25  
3              628          486        7.29  
4              420          621        7.19  
5              247          393        6.30  
6              492          418        5.67  
7              487          392        5.04  
8              498          330        4.22  
9              262          396        3.37  
10              67          184        2.08  
11             125          167        1.87  
12              63          137        1.61  
13              76          126        1.61  
14             123          190        1.23  
15              22           66        1.13  
16              72           79        1.12  
17              55           94        0.92  
18             132           69        0.87  
19              53           78        0.73  
20              53           70        0.54  
21               0           31        0.43  
22              20           30        0.34  
23             123           67         NaN

We used the index_ignore = "true" to avoid messing up the indexing

In [39]:
updated_table.loc[23,"PERCENTAGE"] = round((updated_table.loc[23,"TOTAL_VOTES"]/np.add.reduce(updated_table["TOTAL_VOTES"]))*100,2)
updated_table

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
0           Amazon Web Services (AWS)        33818               29479   
1                     Microsoft Azure        18105               15441   
2                        Google Cloud        16592               13304   
3                            Firebase        10761                8550   
4                          Cloudflare        10599                8326   
5                       Digital Ocean         9299                7838   
6                              Heroku         8358                6535   
7                              Vercel         7431                5727   
8                             Netlify         6228                4708   
9                              VMware         4964                3584   
10                            Hetzner         3069                2465   
11                 Linode, now Akamai         2755                2104   
12                    Managed Hosting         2379                1889   
13                                OVH         2378                1912   
14  Oracle Cloud Infrastructure (OCI)         1810                1146   
15                          OpenShift         1671                1432   
16                             Fly.io         1649                1342   
17                              Vultr         1357                1021   
18                             Render         1287                 920   
19                          OpenStack         1076                 802   
20                IBM Cloud Or Watson          801                 571   
21                           Scaleway          628                 520   
22                         Colocation          497                 385   
23                  My Cloud Platform          500                 323   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  
0              654         1185       22.93  
1              370          714       12.27  
2              652          866       11.25  
3              628          486        7.29  
4              420          621        7.19  
5              247          393        6.30  
6              492          418        5.67  
7              487          392        5.04  
8              498          330        4.22  
9              262          396        3.37  
10              67          184        2.08  
11             125          167        1.87  
12              63          137        1.61  
13              76          126        1.61  
14             123          190        1.23  
15              22           66        1.13  
16              72           79        1.12  
17              55           94        0.92  
18             132           69        0.87  
19              53           78        0.73  
20              53           70        0.54  
21               0           31        0.43  
22              20           30        0.34  
23             123           67        0.34

### Sorting


Like we sorted stuff in numpy we can also sort the table in pandas using the function **.sort_values()** and we can also use arguments like **ascending** and **inplace** which takes up boolean values

*   Ascending argument helps us to determine the sorting order
*   Inplace argument when true means the sorting effect will reflect in the main table





In [40]:
cloud_survey_table_sorted = cloud_survey_table.sort_values("PROFESSIONAL_VOTES",ascending=False,inplace=False)
cloud_survey_table_sorted

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
1           Amazon Web Services (AWS)        33818               29479   
2                     Microsoft Azure        18105               15441   
3                        Google Cloud        16592               13304   
4                            Firebase        10761                8550   
5                          Cloudflare        10599                8326   
6                       Digital Ocean         9299                7838   
7                              Heroku         8358                6535   
8                              Vercel         7431                5727   
9                             Netlify         6228                4708   
10                             VMware         4964                3584   
11                            Hetzner         3069                2465   
12                 Linode, now Akamai         2755                2104   
14                                OVH         2378                1912   
13                    Managed Hosting         2379                1889   
16                          OpenShift         1671                1432   
17                             Fly.io         1649                1342   
15  Oracle Cloud Infrastructure (OCI)         1810                1146   
18                              Vultr         1357                1021   
19                             Render         1287                 920   
20                          OpenStack         1076                 802   
21                IBM Cloud Or Watson          801                 571   
22                           Scaleway          628                 520   
23                         Colocation          497                 385   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  
1              654         1185       22.93  
2              370          714       12.27  
3              652          866       11.25  
4              628          486        7.29  
5              420          621        7.19  
6              247          393        6.30  
7              492          418        5.67  
8              487          392        5.04  
9              498          330        4.22  
10             262          396        3.37  
11              67          184        2.08  
12             125          167        1.87  
14              76          126        1.61  
13              63          137        1.61  
16              22           66        1.13  
17              72           79        1.12  
15             123          190        1.23  
18              55           94        0.92  
19             132           69        0.87  
20              53           78        0.73  
21              53           70        0.54  
22               0           31        0.43  
23              20           30        0.34

### More useful functions

We can have some more basic function which will help us to determine something more about the data

In [59]:
cloud_survey_table_sorted.head() # Gives the top five elements of the table

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  BEGINNER_VOTES  \
1  Amazon Web Services (AWS)        33818               29479             654   
2            Microsoft Azure        18105               15441             370   
3               Google Cloud        16592               13304             652   
4                   Firebase        10761                8550             628   
5                 Cloudflare        10599                8326             420   

   OTHER_VOTES  PERCENTAGE  
1         1185       22.93  
2          714       12.27  
3          866       11.25  
4          486        7.29  
5          621        7.19

In [60]:
cloud_survey_table_sorted.tail() # Gives the bottom five elements of the table

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  BEGINNER_VOTES  \
19               Render         1287                 920             132   
20            OpenStack         1076                 802              53   
21  IBM Cloud Or Watson          801                 571              53   
22             Scaleway          628                 520               0   
23           Colocation          497                 385              20   

    OTHER_VOTES  PERCENTAGE  
19           69        0.87  
20           78        0.73  
21           70        0.54  
22           31        0.43  
23           30        0.34

In [61]:
cloud_survey_table.sum() # Used to caluclate the sum of all columns

CLOUD_PLATFORMS       Amazon Web Services (AWS)Microsoft AzureGoogle...
TOTAL_VOTES                                                      147512
PROFESSIONAL_VOTES                                               120001
BEGINNER_VOTES                                                     5571
OTHER_VOTES                                                        7122
PERCENTAGE                                                       100.01
CATEGORY              PopularPopularPopularPopularPopularPopularPopu...
dtype: object

*   As we can see that the CLOUD_PLATFORMS columns get's a very large string as the sum as it basically doesn't know what to do so it concatenates all the string. (This won't create any issue now but later on we'll find a solution to this problem)
*   Also we can see the sum of all the percentage is 100.01 that's because every device handles precision differently so it must be just close to 100

In [44]:
cloud_survey_table.mean() # Used to calculate the mean of all columns

<ipython-input-44-4db63ed12779>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  cloud_survey_table.mean()


TOTAL_VOTES           6413.565217
PROFESSIONAL_VOTES    5217.434783
BEGINNER_VOTES         242.217391
OTHER_VOTES            309.652174
PERCENTAGE               4.348261
dtype: float64

*   As discussed earlier we can see that the column that has string created an issue while calculating the mean. (Offline compilers will give an error meanwhile google collab doesn't give an error but throws an warning and skips that part of the calculation)
*   So the alternate approach to this could be trying to remove the CLOUD_PLATFORMS column from the table and then try to do the operations on the table



So in order to remove the CLOUD_PLATFORMS column we'll do the following:

1.   Converting the data frame to list
2.   Using the remove() to remove the CLOUD_PLATFORMS column from the list
3.   Creating an copy of the original table and copying it via using the list that we created



In [63]:
# Step-1
columns_without_name = list(cloud_survey_table.columns)

# Step-2
columns_without_name.remove("CLOUD_PLATFORMS")

# Step-3
cloud_survey_table_without_name = cloud_survey_table[columns_without_name].copy()

# Printing
cloud_survey_table_without_name

TOTAL_VOTES  PROFESSIONAL_VOTES  BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE  \
1         33818               29479             654         1185       22.93   
2         18105               15441             370          714       12.27   
3         16592               13304             652          866       11.25   
4         10761                8550             628          486        7.29   
5         10599                8326             420          621        7.19   
6          9299                7838             247          393        6.30   
7          8358                6535             492          418        5.67   
8          7431                5727             487          392        5.04   
9          6228                4708             498          330        4.22   
10         4964                3584             262          396        3.37   
11         3069                2465              67          184        2.08   
12         2755                2104             125          167        1.87   
13         2379                1889              63          137        1.61   
14         2378                1912              76          126        1.61   
15         1810                1146             123          190        1.23   
16         1671                1432              22           66        1.13   
17         1649                1342              72           79        1.12   
18         1357                1021              55           94        0.92   
19         1287                 920             132           69        0.87   
20         1076                 802              53           78        0.73   
21          801                 571              53           70        0.54   
22          628                 520               0           31        0.43   
23          497                 385              20           30        0.34   

     CATEGORY  
1     Popular  
2     Popular  
3     Popular  
4     Popular  
5     Popular  
6     Popular  
7     Popular  
8     Popular  
9    Upcoming  
10   Upcoming  
11   Upcoming  
12   Upcoming  
13   Upcoming  
14   Upcoming  
15   Upcoming  
16  Unpopular  
17  Unpopular  
18  Unpopular  
19  Unpopular  
20  Unpopular  
21  Unpopular  
22  Unpopular  
23  Unpopular

Now we can use functions like mean(), max(), min() etc

In [46]:
cloud_survey_table_without_name.mean()

TOTAL_VOTES           6413.565217
PROFESSIONAL_VOTES    5217.434783
BEGINNER_VOTES         242.217391
OTHER_VOTES            309.652174
PERCENTAGE               4.348261
dtype: float64

In [47]:
cloud_survey_table_without_name.max()

TOTAL_VOTES           33818.00
PROFESSIONAL_VOTES    29479.00
BEGINNER_VOTES          654.00
OTHER_VOTES            1185.00
PERCENTAGE               22.93
dtype: float64

In [48]:
cloud_survey_table_without_name.min()

TOTAL_VOTES           497.00
PROFESSIONAL_VOTES    385.00
BEGINNER_VOTES          0.00
OTHER_VOTES            30.00
PERCENTAGE              0.34
dtype: float64

We can use the function named describe() to get all the important details like count, mean, standard deviation, quartiles, min and max.

In [49]:
cloud_survey_table_without_name.describe()

TOTAL_VOTES  PROFESSIONAL_VOTES  BEGINNER_VOTES  OTHER_VOTES  \
count     23.000000           23.000000       23.000000    23.000000   
mean    6413.565217         5217.434783      242.217391   309.652174   
std     7824.288378         6738.214498      228.147073   301.402270   
min      497.000000          385.000000        0.000000    30.000000   
25%     1503.000000         1083.500000       59.000000    78.500000   
50%     2755.000000         2104.000000      125.000000   184.000000   
75%     8828.500000         7186.500000      453.500000   407.000000   
max    33818.000000        29479.000000      654.000000  1185.000000   

       PERCENTAGE  
count   23.000000  
mean     4.348261  
std      5.304615  
min      0.340000  
25%      1.020000  
50%      1.870000  
75%      5.985000  
max     22.930000

### Categorization

We can categories the data in our table using some condtion and bitwise operators...we'll be using the select() in order to do so

In [50]:
# We'll make sure that it should be in a list and then being add to the table
conditions = [(cloud_survey_table["PERCENTAGE"]>5),((cloud_survey_table["PERCENTAGE"]<=5) & (cloud_survey_table["PERCENTAGE"]>=1.2)),(cloud_survey_table["PERCENTAGE"]<1.2)]
categories = ["Popular","Upcoming","Unpopular"]
cloud_survey_table["CATEGORY"] = np.select(conditions,categories,default="Unknown")
cloud_survey_table

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  \
1           Amazon Web Services (AWS)        33818               29479   
2                     Microsoft Azure        18105               15441   
3                        Google Cloud        16592               13304   
4                            Firebase        10761                8550   
5                          Cloudflare        10599                8326   
6                       Digital Ocean         9299                7838   
7                              Heroku         8358                6535   
8                              Vercel         7431                5727   
9                             Netlify         6228                4708   
10                             VMware         4964                3584   
11                            Hetzner         3069                2465   
12                 Linode, now Akamai         2755                2104   
13                    Managed Hosting         2379                1889   
14                                OVH         2378                1912   
15  Oracle Cloud Infrastructure (OCI)         1810                1146   
16                          OpenShift         1671                1432   
17                             Fly.io         1649                1342   
18                              Vultr         1357                1021   
19                             Render         1287                 920   
20                          OpenStack         1076                 802   
21                IBM Cloud Or Watson          801                 571   
22                           Scaleway          628                 520   
23                         Colocation          497                 385   

    BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE   CATEGORY  
1              654         1185       22.93    Popular  
2              370          714       12.27    Popular  
3              652          866       11.25    Popular  
4              628          486        7.29    Popular  
5              420          621        7.19    Popular  
6              247          393        6.30    Popular  
7              492          418        5.67    Popular  
8              487          392        5.04    Popular  
9              498          330        4.22   Upcoming  
10             262          396        3.37   Upcoming  
11              67          184        2.08   Upcoming  
12             125          167        1.87   Upcoming  
13              63          137        1.61   Upcoming  
14              76          126        1.61   Upcoming  
15             123          190        1.23   Upcoming  
16              22           66        1.13  Unpopular  
17              72           79        1.12  Unpopular  
18              55           94        0.92  Unpopular  
19             132           69        0.87  Unpopular  
20              53           78        0.73  Unpopular  
21              53           70        0.54  Unpopular  
22               0           31        0.43  Unpopular  
23              20           30        0.34  Unpopular

Previously we did calculated stuff column-wise but now as we have created categories we can also calculate stuff category-wise by using the function **groupby()** and using more properties like sum(), mean() etc.

In [51]:
cloud_survey_table.groupby("CATEGORY").sum()

<ipython-input-51-0a424537e85e>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  cloud_survey_table.groupby("CATEGORY").sum()


TOTAL_VOTES  PROFESSIONAL_VOTES  BEGINNER_VOTES  OTHER_VOTES  \
CATEGORY                                                                  
Popular         114963               95200            3950         5075   
Unpopular         8966                6993             407          517   
Upcoming         23583               17808            1214         1530   

           PERCENTAGE  
CATEGORY               
Popular         77.94  
Unpopular        6.08  
Upcoming        15.99

In [52]:
cloud_survey_table.groupby("CATEGORY").mean()

<ipython-input-52-d8ec06c98d5a>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  cloud_survey_table.groupby("CATEGORY").mean()


TOTAL_VOTES  PROFESSIONAL_VOTES  BEGINNER_VOTES  OTHER_VOTES  \
CATEGORY                                                                  
Popular      14370.375           11900.000      493.750000   634.375000   
Unpopular     1120.750             874.125       50.875000    64.625000   
Upcoming      3369.000            2544.000      173.428571   218.571429   

           PERCENTAGE  
CATEGORY               
Popular      9.742500  
Unpopular    0.760000  
Upcoming     2.284286

If we want to use multiple methods we can also use the method aggregate() which helps in finding multiple stuff at the same time

In [53]:
cloud_survey_table.groupby("CATEGORY").aggregate([np.mean,np.median,np.max])

<ipython-input-53-1649fca3fc93>:1: FutureWarning: ['CLOUD_PLATFORMS'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.
  cloud_survey_table.groupby("CATEGORY").aggregate([np.mean,np.median,np.max])


TOTAL_VOTES                 PROFESSIONAL_VOTES                 \
                 mean   median   amax               mean  median   amax   
CATEGORY                                                                  
Popular     14370.375  10680.0  33818          11900.000  8438.0  29479   
Unpopular    1120.750   1181.5   1671            874.125   861.0   1432   
Upcoming     3369.000   2755.0   6228           2544.000  2104.0   4708   

          BEGINNER_VOTES             OTHER_VOTES              PERCENTAGE  \
                    mean median amax        mean median  amax       mean   
CATEGORY                                                                   
Popular       493.750000  489.5  654  634.375000  553.5  1185   9.742500   
Unpopular      50.875000   53.0  132   64.625000   69.5    94   0.760000   
Upcoming      173.428571  123.0  498  218.571429  184.0   396   2.284286   

                         
          median   amax  
CATEGORY                 
Popular     7.24  22.93  
Unpopular   0.80   1.13  
Upcoming    1.87   4.22

### Vectorize string operations

We'll use the following syntax in order to use our normal python string functions with tables

In [54]:
cloud_survey_table["CLOUD_PLATFORMS"].str.startswith("A")

1      True
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
Name: CLOUD_PLATFORMS, dtype: bool

This action will return an boolean array but we can print this using the following syntax

In [55]:
cloud_survey_table[cloud_survey_table["CLOUD_PLATFORMS"].str.startswith("A")]

CLOUD_PLATFORMS  TOTAL_VOTES  PROFESSIONAL_VOTES  BEGINNER_VOTES  \
1  Amazon Web Services (AWS)        33818               29479             654   

   OTHER_VOTES  PERCENTAGE CATEGORY  
1         1185       22.93  Popular

In [56]:
cloud_survey_table["CLOUD_PLATFORMS"].str.upper()

1             AMAZON WEB SERVICES (AWS)
2                       MICROSOFT AZURE
3                          GOOGLE CLOUD
4                              FIREBASE
5                            CLOUDFLARE
6                         DIGITAL OCEAN
7                                HEROKU
8                                VERCEL
9                               NETLIFY
10                               VMWARE
11                              HETZNER
12                   LINODE, NOW AKAMAI
13                      MANAGED HOSTING
14                                  OVH
15    ORACLE CLOUD INFRASTRUCTURE (OCI)
16                            OPENSHIFT
17                               FLY.IO
18                                VULTR
19                               RENDER
20                            OPENSTACK
21                  IBM CLOUD OR WATSON
22                             SCALEWAY
23                           COLOCATION
Name: CLOUD_PLATFORMS, dtype: object

We can also export our data multiple file formats using the following method

In [57]:
cloud_survey_table.to_excel("survey table.xlsx")

We can also read the data from multiple formats

In [58]:
pd.read_excel("/content/survey table.xlsx")

Unnamed: 0                    CLOUD_PLATFORMS  TOTAL_VOTES  \
0            1          Amazon Web Services (AWS)        33818   
1            2                    Microsoft Azure        18105   
2            3                       Google Cloud        16592   
3            4                           Firebase        10761   
4            5                         Cloudflare        10599   
5            6                      Digital Ocean         9299   
6            7                             Heroku         8358   
7            8                             Vercel         7431   
8            9                            Netlify         6228   
9           10                             VMware         4964   
10          11                            Hetzner         3069   
11          12                 Linode, now Akamai         2755   
12          13                    Managed Hosting         2379   
13          14                                OVH         2378   
14          15  Oracle Cloud Infrastructure (OCI)         1810   
15          16                          OpenShift         1671   
16          17                             Fly.io         1649   
17          18                              Vultr         1357   
18          19                             Render         1287   
19          20                          OpenStack         1076   
20          21                IBM Cloud Or Watson          801   
21          22                           Scaleway          628   
22          23                         Colocation          497   

    PROFESSIONAL_VOTES  BEGINNER_VOTES  OTHER_VOTES  PERCENTAGE   CATEGORY  
0                29479             654         1185       22.93    Popular  
1                15441             370          714       12.27    Popular  
2                13304             652          866       11.25    Popular  
3                 8550             628          486        7.29    Popular  
4                 8326             420          621        7.19    Popular  
5                 7838             247          393        6.30    Popular  
6                 6535             492          418        5.67    Popular  
7                 5727             487          392        5.04    Popular  
8                 4708             498          330        4.22   Upcoming  
9                 3584             262          396        3.37   Upcoming  
10                2465              67          184        2.08   Upcoming  
11                2104             125          167        1.87   Upcoming  
12                1889              63          137        1.61   Upcoming  
13                1912              76          126        1.61   Upcoming  
14                1146             123          190        1.23   Upcoming  
15                1432              22           66        1.13  Unpopular  
16                1342              72           79        1.12  Unpopular  
17                1021              55           94        0.92  Unpopular  
18                 920             132           69        0.87  Unpopular  
19                 802              53           78        0.73  Unpopular  
20                 571              53           70        0.54  Unpopular  
21                 520               0           31        0.43  Unpopular  
22                 385              20           30        0.34  Unpopular